### ISSUE TITLE
streaming from an external source

### ISSUE THREAD

**Issue by sdave28jul** [slavavdovichenko/MediaLibDemos3x#27]
Hi,

Can I live stream the video (codec: H264) playing on ipad from an external source using MediaLibDemos3x?

**Comment by slavavdovichenko** [comment]
Yes - see RTMPStreamPlayer demo sample here

**Comment by sdave28jul** [comment]
Thank you for the response.

My exact requirement is as follows:

I have video from a drone camera(which is the external source) playing on my iPad, I want to stream this video stream live from iPad to a media server such as Wowza.

Can I acheive this using MediaLibDemos3x?

**Comment by slavavdovichenko** [comment]
Yes - see RTMPStreamPublisher demo sample ))
In ViewController.m set the url of your wowza - and try:

-(void)viewDidLoad {
    
    //[DebLog setIsActive:YES];
    
    [super viewDidLoad];
    
    socket = nil;
    upstream = nil;

    hostTextField.text = @"rtmp://10.0.1.62:1935/live";
    hostTextField.delegate = self;

    streamTextField.text = @"teststream";
    streamTextField.delegate = self;
}

**Comment by sdave28jul** [comment]
Yes I have tried streaming the iPad camera video to Wowza, however I would like to stream the drone video and not the iPad camera video from iPad to Wowza.

**Comment by slavavdovichenko** [comment]
For your purpose you have to use VIDEO_CUSTOM publisher mode.
Give me your email - I'll send the demo project.

**Comment by sdave28jul** [comment]
Please email at sdave28jul@gmail.com

**Comment by sdave28jul** [comment]
Thank you for the demo project, can you please provide some more pointers on how I could stream the drone video playing on the ipad using this sample?

**Comment by sdave28jul** [comment]
Hi,

I'm not able to stream to server with mode set to VIDEO_CUSTOM.  I tried streaming to wowza with the attached sample but all I see is a black screen on wowza. It would be of great help if you could provide some more pointers.

**Comment by slavavdovichenko** [comment]
Hi,
I am implementing a new sample just now -  I let know when it will be ready

**Comment by sdave28jul** [comment]
Thank you for the response. I already have the H264 video from drone camera playing on my iOS device, hence my requiremnt would be just to stream this video using rtmp and not do any video encoding (as I'm not using the ipad/iphone camera input) 

Please let me know if it is in scope of MediaLibDemos3x to stream any video playing on iOS device (not the iOS camera video) to media server.

**Comment by slavavdovichenko** [comment]
Can you give me the data format that you get from your drone? Can you
represent them in some CoreMedia struct,

CMSampleBuffer or CVImageBuffer, for example?

**Comment by sdave28jul** [comment]
Im receiving the video from a Parrot Bebop drone using the follwing iOS sample

https://github.com/BrennanJones/BebopDroneVideoStreaming-iOS

**Comment by slavavdovichenko** [comment]
Ok,

You can make the following:
1. Add somewhere (maybe as RTMP singleton) the same methods as I've made in SquareCamViewController class of SquareCam sample:
//-------------------------------------- IMediaStreamEvent mehods ------------------------------------------------

@interface SquareCamViewController (IMediaStreamEvent) <MPIMediaStreamEvent>
-(void)connectPublish;
-(void)disconnectPublish;
-(void)startPublish;
-(void)stopPublish;
-(void)sendFrame:(CMSampleBufferRef)sampleBuffer;
@end

@implementation SquareCamViewController (IMediaStreamEvent)

-(void)connectPublish {

    NSLog(@"ConnectPublish\n");    
    if (upstream)
        return;    

    MPMediaPublishType publishType = (NO) ? PUBLISH_LIVE : PUBLISH_RECORD;
    upstream = [[BroadcastStreamClient alloc] init:@"rtmp://192.168.1.100:1935/live"];    
    upstream.delegate = self;
    [upstream setVideoMode:VIDEO_CUSTOM];
    [upstream setAudioMode:AUDIO_OFF];
    [upstream stream:@"iPhoneMedia" publishType:publishType];    
}

-(void)disconnectPublish {   

    NSLog(@"disconnectPublish\n");

    if (!upstream) 
        return;

    [upstream disconnect];
    [upstream release];
    upstream = nil;
}

-(void)startPublish {

    NSLog(@"startPublish\n");

    if (!upstream) 
        return;

    [upstream start];
}

-(void)stopPublish {    

    NSLog(@"stopPublish\n");

    if (!upstream) 
        return;    

    [upstream pause];
}

-(void)sendFrame:(CMSampleBufferRef)sampleBuffer {   

    // Get the frame timestamp

    CMTime presentationTimeStamp = CMSampleBufferGetOutputPresentationTimeStamp(sampleBuffer); 
    int timestamp = presentationTimeStamp.value/(presentationTimeStamp.timescale/1000);

    // Send the frame
CVPixelBufferRef pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer);
    [upstream sendFrame:pixelBuffer timestamp:timestamp];     
}

#pragma mark-

#pragma mark IMediaStreamEvent Methods

-(void)stateChanged:(id)sender state:(int)state description:(NSString *)description {

    NSLog(@" $$$$$$ <IMediaStreamEvent> stateChangedEvent: %d = %@", (int)state, description);    

    switch (state) {            
        case CONN_DISCONNECTED: 
        {
            [self disconnectPublish];            
            break;
        }            
        case CONN_CONNECTED: 
        case STREAM_PAUSED: 
        {
            [self startPublish];
            break;
        }            
        case STREAM_PLAYING: 
        {
            break;
        }
        default:
            break;
    }
}

-(void)connectFailed:(id)sender code:(int)code description:(NSString *)description {

    NSLog(@" $$$$$$ <IMediaStreamEvent> connectFailedEvent: %d = %@\n", code, description);    

    if (code > 0)
        [self disconnectPublish];
}

@end

2. In updateVideoViewWithFrame: method of DroneVideoView class you have 

        CMSampleBufferRef sbRef = NULL;

which is used for frame rendering below:

        dispatch_async(dispatch_get_main_queue(),^{
            [_videoLayer enqueueSampleBuffer:sbRef];
            [_videoLayer setNeedsDisplay];

Thus, sbRef is the structure which you should use for RTMP publishing with sendFrame: method (above)
-(void)sendFrame:(CMSampleBufferRef)sampleBuffer;

Regards,
Slava

**Comment by sdave28jul** [comment]
Hi Slava,

Thankyou, I will try and let you know

**Comment by sdave28jul** [comment]
Hi Slava,

The VIDEO_CUSTOM mode does not seem to work. I have made the changes suggested by you but still I'm neither able to connect nor view the drone video on Wowza with VIDEO_CUSTOM mode. If you have an updated sample, please do share.

**Comment by sdave28jul** [comment]
Hi,

Is there any updated sample with VIDEO_CUSTOM mode successfully working?

Thanks in advance

**Comment by slavavdovichenko** [comment]
Hi,

Yes, I have done the sample app for VIDEO_CUSTOM mode.
You can get it from github -
https://github.com/slavavdovichenko/MediaLibDemos3x (RTMPPhotoPublish
target).
Also you must update the fixed CommLibiOS.a & MediaLibiOS.a from this repo.

Slava

**Comment by sdave28jul** [comment]
Hi,

I'm successfully able to stream to Wowza with the RTMPPhotoPublisher sample.

I have made changes in my project for my requirement to stream drone video to Wowza however I see only a black screen when the stream is playing on wowza.

I have mailed you the changes I have made, can you please check and help out?

Thanks in advance.

**Comment by sdave28jul** [comment]
Follwing up, if there is any help possible for above requirement.

**Comment by slavavdovichenko** [comment]
Hi,
you can try fixed RTMPPhotoPublisher app -> https://github.com/slavavdovichenko/MediaLibDemos3x

**Comment by slavavdovichenko** [comment]
I think you have to investigate this issue - http://stackoverflow.com/questions/29525000/how-to-use-videotoolbox-to-decompress-h-264-video-stream/29525001#29525001
because to integrate the drone stream with the rtmp stream you need the  uncompressed data, thus VTDecompressionSession have to be used.

**Comment by reenaphilip** [comment]
I have tried to decompressesed and tried to stream the CGImageRef. But whats the timeStamp which I have to send along with the CGImageRef

**Comment by vivek1608** [comment]
When i Run And click on connnect then popup come "connectfailedEvent:fault code = -1 < Input or/and Output Stream is not Opened (-1)">
